src/mpid/ch4/src/ch4r_recvq.h

```diff
@@ -158,13 +158,18 @@ MPL_STATIC_INLINE_PREFIX MPIR_Request *MPIDIG_dequeue_posted(int rank, int tag,
     MPIR_T_PVAR_TIMER_START(RECVQ, time_failed_matching_postedq);
     DL_FOREACH_SAFE(*list, curr, tmp) {
         MPIR_T_PVAR_COUNTER_INC(RECVQ, posted_recvq_match_attempts, 1);
-        req = curr->request;
-        if (MPIDIG_match_posted(rank, tag, context_id, req)) {
+#ifndef MPIDI_CH4_DIRECT_NETMOD
+        /* NOTE: extra negation to force logical comparisons */
+        if (!MPIDI_REQUEST(curr->request, is_local) != !is_local) {
```

Comment:
oof. `bool` type would avoid having to do this, no? Also is this related to rndv or just a matching optimization? Rank comparison is already part of matching.

Reply:
Yes, if we use a true boolean type (which C99 provides) then we can avoid this hack. Somehow, I am still nervous that some compilers may still not do the right thing -- i.e. internally use integer and one true may not equal to another true.

Anyway, how about leave this to a separate PR that do the bool refactor in wholesale to maintain consistency?

Reply:
But is this new optimization or a fix? If optimization, then I prefer dropping it from this series.

Reply:
OK, I see we are posting anysource requests twice to the same list with AM enabled 😞